NuRadioReco/modules/analogToDigitalConverter.py

```diff
@@ -254,18 +282,63 @@ def get_digital_trace(self, station, det, channel,
             error_msg += 'Please change the ADC sampling rate.'
             raise ValueError(error_msg)
 
+        # Choosing Nyquist zone
+        if nyquist_zone is not None:
+
+            if nyquist_zone < 1:
+                error_msg = "Nyquist zone is less than one. Exiting."
+                raise ValueError(error_msg)
+            if not isinstance(nyquist_zone, int):
+                try:
+                    nyquist_zone = int(nyquist_zone)
+                except:
+                    error_msg = "Could not convert nyquist_zone to integer. Exiting."
+                    raise ValueError(error_msg)
+
+            passband = ( (nyquist_zone-1) * adc_sampling_frequency/2 + bandwidth_edge,
+                         nyquist_zone * adc_sampling_frequency/2 - bandwidth_edge )
+            filtered_trace = butterworth_filter_trace(trace, input_sampling_frequency,
+                                                      passband)
+        else:
+            filtered_trace = trace[:]
+
+        # Random clock offset
         delayed_times = times + adc_time_delay
-        interpolate_trace = interp1d(times, trace, kind='quadratic',
-                                     fill_value='extrapolate')
+        interpolate_trace = interp1d(times, filtered_trace, kind='quadratic',
```

Comment:
did you check that quadratic interpolation does not create artifacts, e.g., a quick significant over/under shooting of the values? Wouldn't a linear interpolation be better? 
An alternative (and likely saver approach) would be to upsample such that the downsampling factor is an integer. Then you can just take a subselection of the data points. E.g. if the downsampling factor is 2, just take very other sample.

Reply:
Also, I've forgotten to comment something important. I cannot use a standard Fourier resampling method because it removes every component above half the Nyquist frequency, effectively destroying all the information in the higher Nyquist zones.

Reply:
yes, downsampling wouldn't work, but upsampling should be possible, shouldn't it? 
I guess my proposed method would work, but only for discrete time shift. E.g. if you upsample by a factor 10, you could choose the start sample from 0-9 and then take every 10th time bin. 
Maybe/probably the interplation method works better though. If you checked that the quadratic interpolation does not do weird stuff it might be the best. I just saw sometimes a weird behavior with higher order interpolations.

Reply:
I think upsampling followed by interpolation is the way to go. Interpolation causes phase issues at high frequencies and can add power at new frequencies for higher-order polynomials.  In my experience, upsampling by a factor of 4 and then linearly interpolating works decently well although you can of course do some numerical tests. 

If you're just trying to shift every sampled time by the same non-integer amount though, you can also use the Fourier Shift theorem (you just have to be careful around the boundaries since it's cyclic).   Basically a delay of \tau is in the time domain domain corresponds to multiplying by  exp(-j \omega \tau) in the Fourier domain (this is obvious if you think about the definition of group delay). But you'd want to either truncate the signal or add some zeroes (and then remove them) to avoid wraparound problems.  

Actually, unless you're losing a sample, you're going to have a boundary problem here too right?

Reply:
> I think upsampling followed by interpolation is the way to go. Interpolation causes phase issues at high frequencies and can add power at new frequencies for higher-order polynomials. In my experience, upsampling by a factor of 4 and then linearly interpolating works decently well although you can of course do some numerical tests.
> 
> If you're just trying to shift every sampled time by the same non-integer amount though, you can also use the Fourier Shift theorem (you just have to be careful around the boundaries since it's cyclic). Basically a delay of \tau is in the time domain domain corresponds to multiplying by exp(-j \omega \tau) in the Fourier domain (this is obvious if you think about the definition of group delay). But you'd want to either truncate the signal or add some zeroes (and then remove them) to avoid wraparound problems.
> 

Yes, I think that would be better. I'll give it a try.
> Actually, unless you're losing a sample, you're going to have a boundary problem here too right?

That's right. But it's no big deal losing a sample. Or two or three. Voltages are long enough most of the time.

Reply:
I don't think upsample is necessary. The input signals are already at 5 GHz, so I think that's more than enough for an interpolation. It would be nice to make sure that every input signal has a sampling rate of 5 GHz or greater, and if not, upsample the trace.

Reply:
agreed, as long as the sample rate is a few times higher than the highest frequency of interest then there's no need to upsample.

Reply:
Now the delay is carried out using phases in Fourier space and the resulting trace is cropped. I've adjusted a bit the timing in the ADC module but ultimately it's not that relevant because it's just a global offset.